src/pages/Main/Favorites.jsx

```diff
@@ -0,0 +1,114 @@
+} from "../../functions/exportFunctions";
+import {LOGIN} from "../../App";
+import {QuizzesContainer} from "./Main";
+
+export const Favorites = ({
+                              changeFilter,
+                              activeFilters,
+                              searchValue,
+                          }) => {
+    const { auth } = useAuth();
+    const [loading, setLoading] = useState(true);
+    const [error, setError] = useState(null);
+    const [favoriteQuizzes, setFavoriteQuizzes] = useState([]);
+    const [favoritesIds, setFavoritesIds] = useState([]);
+    const [initialDataLoaded, setInitialDataLoaded] = useState(false);
+    const [userQuizzesUpdated, setUserQuizzesUpdated] = useState(false);
+    const navigate = useNavigate();
+
+    const removeFavoriteHandler = async (quizId) => {
+        try {
+            await removeFavorite(quizId);
+            setFavoritesIds((prevFavoriteQuizzesIds) =>
+                prevFavoriteQuizzesIds.filter((favorite) => favorite !== quizId)
+            );
+            setFavoriteQuizzes((prevFavoriteQuizzes) =>
+                prevFavoriteQuizzes.filter((quiz) => quiz.id !== quizId)
+            );
+        } catch (error) {
+            console.error('Error removing favorite:', error);
+        }
+    };
+
+    useEffect(() => {
+        const fetchAndSetFavoriteQuizzes = async () => {
+            try {
+                setLoading(true);
+                const quizzesIds = await getFavorites();
+                setFavoritesIds(quizzesIds);
+                if (auth.loggedIn && favoriteQuizzes.length === 0 && !initialDataLoaded) {
+                    console.log('fetchFavorites')
+                    await fetchFavorites(
+                        backendApiCall,
+                        setFavoriteQuizzes,
+                        setError,
+                        auth,
+                        setLoading
+                    );
+                    setInitialDataLoaded(true);
+                } else if (!auth.loggedIn) {
+                    navigate(LOGIN);
+                } else {
+                    setLoading(false);
+                }
+            } catch (error) {
+                setError(error);
+            }
+        };
+        fetchAndSetFavoriteQuizzes();
+    }, [auth.loggedIn, setFavoritesIds, auth, favoriteQuizzes.length, navigate]);
+
+
+    useEffect(() => {
+        const fetchQuizzes = async () => {
+            try {
+                if (auth.loggedIn &&
+                    initialDataLoaded &&
+                    favoriteQuizzes.length > 0 &&
+                    !userQuizzesUpdated) {
+                    await fetchFavoritesAndAddUserQuizzes(
+                        backendApiCall,
+                        favoriteQuizzes,
+                        setFavoriteQuizzes,
+                        setError,
+                        auth,
+                        setUserQuizzesUpdated
+                    );
+                } else if (!auth.loggedIn) {
```

Comment:
Good idea to check if the user is logged in!

This is not something you need to change, but just for your future reference:

I find it easiest to manage login redirects by using a ProtectedRoute component like the following. The ProtectedRoute component checks that there's a valid session token (either by loading a cookie or by making a request), and rendering LoggedInApp if there's a valid token or redirecting to the Login page if not. With this setup, every component you add under LoggedInApp will be behind an auth wall and you won't have to worry about whether the user is logged in.

```jsx
        <Router>
            <Switch>
                <Route path={PATHS.LOGIN} component={Login} />
                <Route path={PATHS.RESET_PASSWORD} component={ResetPassword} />
                <Route
                    path={PATHS.REQUEST_RESET_PASSWORD}
                    component={RequestResetPassword}
                />
                <ProtectedRoute path={PATHS.HOME} component={LoggedInApp} />
            </Switch>
        </Router>
```

Reply:
Great idea, you're right, I've seen this on projects before! Thank you